### DISCUSSION TITLE
Functions: share your use-cases

### DISCUSSION THREAD

**Discussion by kiwicopple** [supabase/supabase#4269] [Feature Requests]
As we start [designing Supabase Functions](https://github.com/supabase/rfcs/blob/rfc/functions/rfc/0004-functions.md), we want to make sure the developer experience we deliver is world-class. To achieve this, we need your feedback.

You can help us in one of two ways:

- Comments
- Upvotes

## Comments

Comment in this thread with your use-case for Functions. Specifically we want to know:

- What use-cases do you have for Functions? What would you use Supabase Functions for?
- What's preventing you from using PostgreSQL Functions?
- Why do you want Supabase Functions, especially when there are already some great options out there (like Vercel and Netlify)

We don't want to know (and will hide/delete), how you would _implement_ functions. We only want to hear your pain points, and from there we can collectively come up with the best solution in this [RFC](https://github.com/supabase/rfcs/blob/rfc/functions/rfc/0004-functions.md).

## Upvotes

If a comment matches the problems _you_ currently face, just upvote it! You don' need to add a duplicate comment - we will be able to gauge from upvotes and comments.

**Comment by bohan0** [comment]
Processing payments with payment providers like Stripe/Square. On the client side, the the cards details are sent directly to Stripe, which returns a token the is stored in Db.

On the server side, a function needs to be written that calls the Stripe charge API using a secret key and the token to actually process the payment.

This could in theory be done with plv8 functions, but not sure if this is best practice. Also there are issues with loading all the right dependences (and updating these - Stripe's js libraries are designed to be managed via npm)

**Reply by prescience-data** [reply]
I have done this with raw Postgres Functions and although it technically works it's a nightmare if anyone needed to try and understand wtf is going on (as it's such a "not for purpose" use of the the `http` plugin, and an arcane / non-standard custom design pattern). It also requires storing the Supabase key in an `env` table which although probably not really an issue if you lock it down properly, still gives me the creeps. 

This is for sure the perfect use of `Supabase Functions` (we are moving all the logic to Cloudflare Workers but having it all on-platform would be amazing).

**Comment by roker15** [comment]
I see a ciritical issue in making query with supabase is **`ambiguity`** when i have to refer foreign kyes. Even if you don't touch your current tables and Just create a **`view`**. It might break all your api call in your application. Never faced this kind of problem when was using hibernate in java. What u people doing about this. This is i think most critical thing for u guys. Just one change in database and everything is borken.

**Reply by liaujianjie** [reply]
Doesn't [generated types](https://supabase.com/docs/reference/javascript/generating-types) solve your issue with this?

**Reply by steve-chavez** [reply]
This would fit better as an issue on https://github.com/supabase/postgrest-js/issues

**Edit**: Nvm, I see https://github.com/supabase/supabase/issues/4291 is already open.

**Comment by liaujianjie** [comment]
### Why not Postgres functions?
- Testing tooling, libraries and frameworks for JS is much more well-known than the same for Postgres.
- I could be wrong about this: Postgres functions aren't able to make API calls to 3rd party APIs. Even if it were possible, I wouldn't want to store my secret keys in the database. What would the DX look like if I tried to use official client libraries in Postgres functions? I imagine it isn't the most ideal.
  - Even if those weren't an issue and we could use PLV8, how can we extend the same DX we have in regular JS/TS to PLV8 functions? Example: How can we have Prettier and ESLint support in the PLV8 functions? Even if it were possible, do we want to assume the overhead of managing this additional config?

### What kind of functions do we have right now?
1. Functions that integrate with 3rd party APIs.
   - Example: endpoint for creating a Stripe checkout session
   - Example: endpoint for sending welcome email to new users via SendGrid
   - Example: endpoint for image processing workloads
2. Functions that receive webhook events
   - Example: endpoint for Stripe webhook event to enable paid features when a paid subscription is created

**Comment by keybits** [comment]
- **Use case:** Data loader to allow users of an app to import data (e.g. 50MB JSON file and parse into normalized relations)
- **Why not PostgreSQL Functions:** Need a place to temporarily store the uploaded file. The data load could take some time, so queuing might be needed.
- **Why Supabase functions?** Payload, CPU and memory limits not enough on existing serverless function solutions.

(Note: I'm not super experienced in this domain, so it might be there's potential for using object storage and PostgreSQL functions for this.)

**Comment by miketromba** [comment]
### What use-cases do you have for Functions? What would you use Supabase Functions for?
For most serious, production apps that I build, I find it way more convenient / intuitive / less prone to vulnerability to have a custom access layer between my users and my data. The data access role-level-security / permissions functionality is ok, but it just doesn't compete with the level of validation / authorization you can do on an authoritative node server... Functions would save me from having to host my own express API and condense my cloud footprint so that my whole app can run within the Supabase ecosystem. This also saves me from having to learn how to build complex Postgres policies.

### What's preventing you from using PostgreSQL Functions?
The dev experience of writing JavaScript functions and leveraging the huge ecosystem of tooling / packages doesn't compare. I already have strong competency in all things JavaScript, and would rather not sink time into learning postgres functions.

### Why do you want Supabase Functions, especially when there are already some great options out there (like Vercel and Netlify)
A) A condensed cloud footprint is attractive - aka. not having to manage accounts with various cloud providers to power a single app.
B) If there is some way to run the Supabase functions in the same machine / data center of the database, could there be significant latency benefits? E.g. hit function -> query for data which lives on same machine/same data center with 0 server-to-database network latency.

**Reply by gabriellemon** [reply]
I agree with you on this one. Having less context and cloud provider switching sounds great.

**Comment by prescience-data** [comment]
## What use-cases do you have for Functions? What would you use Supabase Functions for?
- For more complex code than is practical in native Postgres Functions, and interacting with third-party APIs / async tasks such as authenticated webhook listeners.

## What's preventing you from using PostgreSQL Functions?

- Nothing, for the most part, this is what we are currently implementing. The issue is that a common language (TypeScript) means devs with less proficiency in PGSQL can immediately be repurposed to safely write functions. 
- I imagine, however, we would continue to primarily use native Postgres functions for any performance benefit on simple things, and use Supabase Functions for more complex code and interacting with third-party APIs.

## Why do you want Supabase Functions, especially when there are already some great options out there (like Vercel and Netlify)

- Reducing the number of "moving parts" is the most attractive thing about Supabase in general.  Although naturally these other services exist, they introduce unnecessary complexity and "crossing cloud" latency.
- Having new devs immediately be able to "see" the full application scope all in one place provides genuine hard dollar savings in hourly rates, as well as reducing the devsec risk of forgetting to disable one of 500 logins (if not using SSO)

**Comment by Electric-Dragon** [comment]
### What use-cases do you have for Functions? What would you use Supabase Functions for?

- ML predictions in a serverless environment (packaging code in a docker container and running it as a function)

### What's preventing you from using PostgreSQL Functions?

- Not sure if PostresSQL functions support docker-based functions.

**Reply by magick93** [reply]
Can launch docker containers using [Postgres Container Apps](https://blog.crunchydata.com/blog/announcing-postgres-container-apps-easy-deploy-postgres-apps).

**Comment by Zegoverno** [comment]
### What use-cases do you have for Functions? What would you use Supabase Functions for?
I have dynamic columns that should be updated daily on end-of-day. would be awesome to have that managed inside the database
### What's preventing you from using PostgreSQL Functions?
Knowledge and experience
### Why do you want Supabase Functions, especially when there are already some great options out there (like Vercel and Netlify)
Reduce number of moving parts

**Comment by riderx** [comment]
Have you check what firebase does with they marketplace function ?

If you achieve to do something like this it could solve a lot of problem from the ground !
https://firebase.google.com/products/extensions?authuser=0&hl=en

**Comment by chasers** [comment]
Not sure how relevant this is in 2022 but pixel trackers usually do a GET request and use the url params for the data they save.

And what about link shorteners?

**Comment by mlanza** [comment]
# What use-cases do you have for Functions? What would you use Supabase Functions for?

I am working on site to facilitate playing turn-based board games similar to Board Game Arena or Yucata.  I contained the entire server-side brains for a game to a single function (with essentially 2 routes, 1 for viewing the state as of a point in time and 1 for issuing a command to progress the present state).

To accomplish this I need some tools, so I use a library which just provides utility functions similar to what's provided in Clojure.  Each game is written using a functional core where a game session is a reduction over an event log.  Events are derived from user and system issued commands.  It's basically event sourcing with a functional core (on the server) and imperative shell (on the client).

# What's preventing you from using PostgreSQL Functions?
Initially I tried to use a CloudFlare Worker, but its free-tier limitations couldn't handle a fat function.  Basically, my function has all the logic for reducing events and transforming newly entered commands into events.  The function returns those events so they can be persisted to the database.

Since the Worker couldn't handle it, I eventually realized I could tuck this program into a JavaScript PostgresSQL Function.  Initially I implemented a card game for its simplicity.  Then I began implementing an actual eurogame with a bit more complexity.  Everything was working until I progressed a game (implementing as I needed to make moves) to about 225 discrete events.

The initial model was kept super simple.  I basically query the database for all events so far and reduce over them each time someone needs to view one frame of the game.  It's turn-based, not real time mind you.  It also does this whenever it needs to determine whether a new command is valid.  And since the user can navigate forwards and backward over the game to see what's happened you can imagine this reduction can become increasingly expensive.  Fortunately, I use HTTP headers to enable caching so that the DB doesn't need to be hit except once per frame per client.

Right now I am working on overcoming CPU time and memory issues.  Eventually, at a certain move Postgres runs out of memory.  So I moved the function logic from the database to an edge function.  I also minified the program.  To be clear, since I needed the library I was already looking at something around the size of a jQuery library.  Then to that you add a full functional core program which can process events/commands, minified about 151KB, so not huge.  But when it is handed 225 events in later frames for a game about 80% complete, it adds up.  I am looking at caching snapshots to mitigate having to reconstitute the game state in its entirety on each frame.

# Why do you want Supabase Functions, especially when there are already some great options out there
I do like that everything exists under one cohesive umbrella and that it abstracts away some of the security issues which are otherwise required to integrate with other vendors.

# What would help?
Since serverless is a relatively new space there aren't a lot of examples demonstrating how to take advantage of the serverless architecture given its constraints.  For example, AWS is sometimes called serverless but it allows functions to be uploaded into containers.  Its functions are 150MB whereas mine are around 150KB.  So calling these serverless is of no help.  They're two different animals.

Plus there's a bit of a disagreement about using monolithic edge functions or many discrete edge functions.  In my case monolith makes sense, but there are a lot of limitations I've been battling with.

Part of the desire to move to serverless is to abstract away hardware issues, but in my experience I'm still dealing with them even if their scope has narrowed.  I feel like this is still a tech space for pioneers.  It would be great if more individuals shared the hurdles in their designs and how they overcame them.
